content/AlgorytmyWykladniczeZadania.tex

```diff
@@ -80,6 +80,89 @@ \subsection{Zadanie 1 -- Najmniejszy zbiór dominujący}
+Zaprojektuj jak najszybszy algorytm rozwiązujący problem 3-SAT i oszacuj jego złożoność.
+
+\paragraph{Rozwiązanie: } Rozważmy algorytm naiwny (używający \textit{bruteforce}'a). Będziemy brać zmienną $x$, która jeszcze nie ma wartościowania i przypisywać jej wartość \true{} lub \false; a następnie rekurencyjnie wywoływać funkcję dla kolejnych zmiennych. Na każdym poziomie rekurencji liniowo sprawdzamy, czy wszystkie klauzule zostały spełnione lub któraś jest niespełniona. Opisany algorytm działa w czasie $O^*(2^n)$ (liczba wszystkich ciągów binarnych długości $n$).
+
+Algorytm naiwny możemy ulepszyć, zauważając następujący fakt: do spełnienia klauzuli wystarczy, żeby co najmniej jeden z literałów miał wartość \textit{true}. Dodatkowo zmienne będziemy rozpatrywać w kolejności pojawiania się w jeszcze nierozstrzygniętych klauzulach.
+
+Mając nierozstrzygniętą klauzulę $\{x, y, z\}$, w której żadna zmienna nie ma przypisanej wartości, będziemy rozpatrywać następujące wartościowania\daggerfootnote{Naturalnie przypadków będzie mniej, jeżeli klauzula ma mniej literałów lub któraś ze zmiennych ma już przypisaną wartość -- wtedy rozważamy odpowiednio jeden lub dwa pierwsze przypadki.}:
+
+\begin{align*}
+	\begin{cases}
+		x = 1;
+	\end{cases}
+	\qquad
+	\begin{cases}
+		x = 0, \\
+		y = 1;
+	\end{cases}
+	\qquad
+	\begin{cases}
+		x = 0, \\
+		y = 0,\\
+		z = 1.
+	\end{cases}
+\end{align*}
+
+W ten sposób maksymalnie osiem możliwych przypadków wartościowania zmiennych w klauzuli $\{x, y, z\}$ ograniczyliśmy do trzech. 
+
+\begin{algorithm}[H]
+	\caption{Rozwiązanie problemu 3-SAT}
+	\begin{algorithmic}[1]
+		\Procedure{Solve3SAT}{$\mathfrak{K}$: zbiór klauzul nierozstrzygniętych}
+		% \If {przy obecnym wartościowaniu któraś z klauzul z $\mathfrak{K}$ jest niespełnialna}
+		% 	\State \Return \false
+		% \EndIf
+		\If {$\mathfrak{K} = \emptyset$}
+		\State \Return \true
+		\EndIf
+		
+		\State 
+		\State \textit{Kasia} $\gets$ nierozstrzygnięta klauzula z $\mathfrak{K}$ 
+		\If{\textit{Kasia} jest niespełnialna przy obecnym wartościowaniu}
+		\State \Return \false
+		\EndIf
+		\State $x, y, z \gets $ literały \textit{Kasi} bez wartościowania
+		\State
+		\State $x \gets $ \true
+		\State $\mathfrak{K}' \gets \mathfrak{K}$ bez \textit{Kasi} i klauzul spełnionych przez $x$
+		\If {\textsc{Solve3SAT}($\mathfrak{K}'$)}
+		\State \Return \true
+		\EndIf
+		\State
+		\State $x \gets $ \false, $y \gets $ \true \Comment{Jeśli literałów było $\geq 2$}
+		\State $\mathfrak{K}' \gets \mathfrak{K}$ bez \textit{Kasi} i klauzul spełnionych przez $y$
+		\If {\textsc{Solve3SAT}($\mathfrak{K}'$)} 
+		\State \Return \true
+		\EndIf
+		\State
+		\State $x \gets $ \false, $y \gets $ \false, $z \gets $ \true \Comment{Jeśli literałów było 3}
+		\State $\mathfrak{K}' \gets \mathfrak{K}$ bez \textit{Kasi} i klauzul spełnionych przez $z$
+		\If {\textsc{Solve3SAT}($\mathfrak{K}'$)}
+		\State \Return \true

```

Comment:
to trochę ogólna uwaga, czy jeśli ustawiamy x=false, y=true, to nie odrzucamy klauzul spełnionych przez $y$ ORAZ spełnionych przez $~x$?

Reply:
moim zdaniem można odrzucic tez te klauzule, ale to nie powinno zmienic oszacowania złozoności, bo złożoność zależy od liczby zmiennych a nie klauzul